test/test_autograd.py

```diff
@@ -4195,6 +4198,26 @@ def fn3(x):  # C -> R
             gradcheck(fn3, (x_c,))
         self.assertFalse(gradcheck(fn3, (x_c,), raise_exception=False))
 
+    def test_gradcheck_dense_and_sparse_inputs(self):
+        def fn(x, y):
+            return x * y.coalesce().to_dense()
+        a = torch.rand(2, 2, requires_grad=True)
+        b = torch.rand(2, 2).to_sparse().requires_grad_(True)
+        self.assertTrue(gradcheck(fn, (a, b), check_sparse_nnz=True, check_batched_grad=False))
+
+    @unittest.skipIf(not torch._C.has_mkldnn, "MKL-DNN build is disabled")
+    def test_gradcheck_multiple_mkldnn_inputs(self):
+        def fn(x, y):
+            return x + y.to_dense()
+        a = torch.rand(10, requires_grad=True)
+        b = torch.rand(10, dtype=torch.float32).to_mkldnn().requires_grad_(True)
+        self.assertTrue(gradcheck(fn, (a, b), atol=1e-1, check_batched_grad=False))
+
+        def fn2(x, y):
+            return x.to_dense() + y.to_dense()
+        c = torch.rand(10, dtype=torch.float32).to_mkldnn().requires_grad_(True)
+        self.assertTrue(gradcheck(fn, (a, c), atol=1e-1, check_batched_grad=False))
```

Comment:
These atols are very lenient, is that expected?

Reply:
> > This PR addresses the comment here #53857 (comment), to reduce the run-time of old gradcheck's get numerical jacobian by a factor of num_outputs. However, because very few ops actually return multiple outputs, there is not too much real speed up here.
> 
> What exactly changed? Does this PR make it so that `func` only gets called 2 * num_inputs times?
> 
> > However, because very few ops actually return multiple outputs
> 
> Are we talking about number of output tensors, or the total numel of all output tensors?

Oh I think there was a newline in the original comment that cut the last part off. A more precise way of saying it should be: New: `2 * total_number_of_elements_in_all_the_inputs * num_outputs`
Old: `2 * total_number_of_elements_in_all_the_inputs`
Fast gradcheck: `2`

This is because before we had this line here https://github.com/pytorch/pytorch/pull/54378/files#diff-440fd3cf8800df054b6c18370faac5566e0ee1f14cc6c8d93d6f436b37dfabc3L302
where we restrict ourselves to utilize only a single output at a time. The way it was done previously is that we looped through both outputs and inputs in a doubly nested loop, and inside this we compute the Jacobian corresponding to input i and output j, which required calling f `2 * `input_i.numel` times. 

After this PR, we utilize the entire potential of calling `f`, so we only loop through the inputs. And since calling f returns all the outputs, we can compute all Jacobians corresponding to input i with the same 2 * `input_i.numel` calls to f.

Reply:
Oh wait I just checked and the actual and expected actually differ by 1.33e-2, so we'd indeed need an atol of 1e-1 (or at least 2e-2).